tikv-client/src/main/java/com/pingcap/tikv/region/RegionStoreClient.java

```diff
@@ -131,35 +152,44 @@ public ByteString get(BackOffer backOffer, ByteString key, long version) {
 
       GetResponse resp = callWithRetry(backOffer, TikvGrpc.METHOD_KV_GET, factory, handler);
 
-      if (resp.hasRegionError()) {
-        backOffer.doBackOff(BoRegionMiss, new RegionException(resp.getRegionError()));
-        continue;
+      if (getNeedRetry(backOffer, resp)) {
+        return resp.getValue();
       }
+    }
+  }
 
-      if (resp.hasError()) {
-        if (resp.getError().hasLocked()) {
-          Lock lock = new Lock(resp.getError().getLocked());
-          boolean ok =
-              lockResolverClient.resolveLocks(backOffer, new ArrayList<>(Arrays.asList(lock)));
-          if (!ok) {
-            // if not resolve all locks, we wait and retry
-            backOffer.doBackOff(
-                BoTxnLockFast, new KeyException((resp.getError().getLocked().toString())));
-          }
-
-          continue;
-        } else {
-          // retriable or abort
-          // this should trigger Spark to retry the txn
-          throw new KeyException(resp.getError());
+  private boolean getNeedRetry(BackOffer backOffer, GetResponse resp) {
+    if (resp == null) {
+      this.regionManager.onRequestFail(region);
+      throw new TiClientInternalException("GetResponse failed without a cause");
+    }
+
+    if (resp.hasRegionError()) {
```

Comment:
This is not needed. It will be dealt with by KVErrorHandler.

Reply:
if key's region changes, should we throw an error and do retry in caller's logic?

Reply:
@birdstorm what do u think?

Reply:
we should. CallWithRetry does not retry when region changes and depends on caller's logic to retry.